File: solutions/java/bank-account/1/src/main/java/BankAccount.java
class BankAccount {
    private boolean isOpen;
    private int balance;
    
    public BankAccount(){
        this.isOpen = false;
        this.balance = 0;
    }

    public void open() {
        this.isOpen = true;
    }

    public void close() {
        this.isOpen = false;
    }

    public synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        if (!this.isOpen) 
            throw new BankAccountActionInvalidException("Account closed");
        if (amount < 0) 
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");

        this.balance += amount;
    }

    public synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        if (!this.isOpen) 
            throw new BankAccountActionInvalidException("Account closed");
        if (this.balance == 0) 
            throw new BankAccountActionInvalidException("Cannot withdraw money from an empty account");
        if (amount < 0) 
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
        if (amount > this.balance) 
            throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");

        this.balance -= amount;
    }

    public int getBalance() throws BankAccountActionInvalidException {
        if (!this.isOpen) throw new BankAccountActionInvalidException("Account closed");

        return this.balance;
    }
}

File: solutions/javascript/bank-account/1/bank-account.js
export class BankAccount {
  #balance = 0;
  #status = "closed";

  open() {
    if (this.#status == "open") throw new ValueError("Account is already open");

    this.#status = "open";
  }

  close() {
    if (this.#status == "closed")
      throw new ValueError("Account is already closed");

    this.#balance = 0;
    this.#status = "closed";
  }

  deposit(amount) {
    if (amount < 0) throw new ValueError("cannot deposit negative amount");
    if (this.#status == "closed") throw new ValueError("Account is closed");
    
    this.#balance += amount;
  }

  withdraw(amount) {
    if (this.#status == "closed") throw new ValueError("Account is closed");
    if (amount < 0) throw new ValueError("cannot withdraw negative amount");
    if (amount > this.#balance)
      throw new ValueError("cannot withdraw more than deposited");

    this.#balance -= amount;
  }

  get balance() {
    if (this.#status == "closed") throw new ValueError("Account is closed");

    return this.#balance;
  }
}

export class ValueError extends Error {
  constructor() {
    super("Bank account error");
  }
}
